source/common/router/config_impl.h

```diff
@@ -69,7 +70,56 @@ class PerFilterConfigs {
 };
 
 class RouteEntryImplBase;
+class RouteMatchContext;
+class VirtualHostImpl;
+
 using RouteEntryImplBaseConstSharedPtr = std::shared_ptr<const RouteEntryImplBase>;
+using RouteEntryImplBaseConstSharedPtrItr =
+    std::vector<RouteEntryImplBaseConstSharedPtr>::const_iterator;
+using RouteEntryImplBaseConstSharedPtrItrOptional =
+    absl::optional<RouteEntryImplBaseConstSharedPtrItr>;
+
+class RouteMatchContext {
```

Comment:
Why is this context and all of the allocations needed? Naively it seems like we would be able to have our route match condition also include calling the callback (if one is provided) and that should be enough?

Reply:
If the callback repeatedly returned continue, I didn't repeatedly identify the virtual host amd loop through the routes to skip over until the last match, context allows me to remember the iterator position and virtual host.

Reply:
The virtual host match happens first, then the route match. AFAIK we shouldn't ever match more than one vhost, so the acceptance happens within a vhost. What am I missing?

Reply:
That's correct matt but since we thread the call all the way through unless I put this information away somewhere lookup will repeat, flow is like this

    https://github.com/envoyproxy/envoy/blob/master/source/common/router/config_impl.h#L885 (this repeatedly call route as long as the callback return continue) ->  
        https://github.com/envoyproxy/envoy/blob/master/source/common/router/config_impl.cc#L1175 (this does vhost lookup - if I don't pass along the vhost from the above lookup will  be repeated) ->
           https://github.com/envoyproxy/envoy/blob/master/source/common/router/config_impl.cc#L1106 (this iterates through the routes unless I pass along the previous iterator location it will have to loop through and skip)

Reply:
Why is threading this through this loop not sufficient? https://github.com/envoyproxy/envoy/blob/f84440dc4f95890f14e2e0686b07258f030b54b3/source/common/router/config_impl.cc#L1126-L1131